File: payment-receiver/domain/payment_event.go
package domain

import (
	"errors"
	"time"
)

// PaymentStatus represents allowed payment statuses
const (
	StatusPaid     = "paid"
	StatusRefunded = "refunded"
)

var validStatuses = map[string]struct{}{
	StatusPaid:     {},
	StatusRefunded: {},
}

func IsValidStatus(status string) bool {
	_, ok := validStatuses[status]
	return ok
}

// PaymentEvent represents a domain entity for a payment webhook
type PaymentEvent struct {
	ID         string
	Amount     int
	Currency   string
	Method     string
	Status     string
	OccurredAt time.Time
}

// NewPaymentEvent creates a validated PaymentEvent entity
func NewPaymentEvent(
	id string,
	amount int,
	currency string,
	method string,
	status string,
	occurredAt string,
) (*PaymentEvent, error) {
	if id == "" || amount == 0 || currency == "" || method == "" || status == "" || occurredAt == "" {
		return nil, errors.New("all fields are required")
	}

	if !IsValidStatus(status) {
		return nil, errors.New("invalid status")
	}

	ts, err := time.Parse(time.RFC3339, occurredAt)
	if err != nil {
		return nil, errors.New("invalid occurred_at format")
	}

	return &PaymentEvent{
		ID:         id,
		Amount:     amount,
		Currency:   currency,
		Method:     method,
		Status:     status,
		OccurredAt: ts,
	}, nil
}

File: payment-processor/src/PaymentProcessor/Entities/PaymentEvent.cs
namespace PaymentProcessor.Domain.Entities;

public class PaymentEvent
{
    public string Id { get; set; } = default!;
    public int Amount { get; set; }
    public string Currency { get; set; } = default!;
    public string Method { get; set; } = default!;
    public string Status { get; set; } = default!;
    public DateTime OccurredAt { get; set; }

    public PaymentEvent(string id, int amount, string currency, string method, string status, DateTime occurredAt)
    {
        if (string.IsNullOrWhiteSpace(id))
            throw new ArgumentException("Id is required.", nameof(id));

        if (amount <= 0)
            throw new ArgumentException("Amount must be greater than 0.", nameof(amount));

        if (string.IsNullOrWhiteSpace(currency))
            throw new ArgumentException("Currency is required.", nameof(currency));

        if (string.IsNullOrWhiteSpace(method))
            throw new ArgumentException("Method is required.", nameof(method));

        if (!IsValidStatus(status))
            throw new ArgumentException($"Invalid status: {status}", nameof(status));

        Id = id;
        Amount = amount;
        Currency = currency;
        Method = method;
        Status = status;
        OccurredAt = occurredAt;
    }

    private bool IsValidStatus(string status)
    {
        // Define valid statuses here
        return new[] { "paid", "failed", "cancelled" }.Contains(status.ToLowerInvariant());
    }
}